providers/gce/gce_loadbalancer_external.go

```diff
@@ -112,45 +112,152 @@ func (g *Cloud) ensureExternalLoadBalancer(clusterName string, clusterID string,
 		g.deleteWrongNetworkTieredResources(loadBalancerName, lbRefStr, netTier)
 	}
 
-	// Check if the forwarding rule exists, and if so, what its IP is.
-	fwdRuleExists, fwdRuleNeedsUpdate, fwdRuleIP, err := g.forwardingRuleNeedsUpdate(loadBalancerName, g.region, requestedIP, ports)
-	if err != nil {
-		return nil, err
-	}
-	if !fwdRuleExists {
-		klog.V(2).Infof("ensureExternalLoadBalancer(%s): Forwarding rule %v doesn't exist.", lbRefStr, loadBalancerName)
+	// If the feature gate is not enabled, we check for multiple protocols and error out.
+	if !g.AlphaFeatureGate.Enabled(AlphaFeatureMultiProtocolLB) {
+		fwdRuleExists, fwdRuleNeedsUpdate, fwdRuleIP, err := g.forwardingRuleNeedsUpdate(loadBalancerName, g.region, requestedIP, ports)
+		if err != nil {
+			return nil, err
+		}
+		if !fwdRuleExists {
+			klog.V(2).Infof("ensureExternalLoadBalancer(%s): Forwarding rule %v doesn't exist.", lbRefStr, loadBalancerName)
+		}
+		// Single-protocol logic
+		ipAddressToUse, isSafeToReleaseIP, err := g.ensureIPAddress(loadBalancerName, lbRefStr, apiService.Spec.LoadBalancerIP, fwdRuleIP, netTier)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if isSafeToReleaseIP {
+				if err := g.DeleteRegionAddress(loadBalancerName, g.region); err != nil && !isNotFound(err) {
+					klog.Errorf("ensureExternalLoadBalancer(%s): Failed to release static IP %s in region %v: %v.", lbRefStr, ipAddressToUse, g.region, err)
+				} else if isNotFound(err) {
+					klog.V(2).Infof("ensureExternalLoadBalancer(%s): IP address %s is not reserved.", lbRefStr, ipAddressToUse)
+				} else {
+					klog.Infof("ensureExternalLoadBalancer(%s): Released static IP %s.", lbRefStr, ipAddressToUse)
+				}
+			} else {
+				klog.Warningf("ensureExternalLoadBalancer(%s): Orphaning static IP %s in region %v: %v.", lbRefStr, ipAddressToUse, g.region, err)
+			}
+		}()
+
+		sourceRanges, err := servicehelpers.GetLoadBalancerSourceRanges(apiService)
+		if err != nil {
+			return nil, err
+		}
+
```

Comment:
this logic is super complex that it is better to use unit tests to avoid issues

Reply:
Definitely - I'm planning on breaking this down into manageable steps, and I think most of those will be about making the code simpler.